cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -246,6 +278,52 @@ public void run() {
     }
   }
 
+  private TransactionContext getTransactionContextFromContext(WorkflowContext context) {
+    TransactionContext transactionContext = new TransactionContext(txClient);
+    Collection<Dataset> datasets = ((BasicWorkflowContext) context).getDatasets().values();
+    for (Dataset ds : datasets) {
+      transactionContext.addTransactionAware((TransactionAware) ds);
+    }
+    return transactionContext;
+  }
+
+  private void runInTransaction(WorkflowAction action, WorkflowContext workflowContext) {
+    TransactionContext transactionContext = getTransactionContextFromContext(workflowContext);
+    try {
+      transactionContext.start();
+      action.run();
```

Comment:
You need to abort the transaction context if the user method call throws exception. See how https://github.com/caskdata/tephra/blob/develop/tephra-core/src/main/java/co/cask/tephra/DefaultTransactionExecutor.java#L131 is implemented.